optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java

```diff
@@ -121,66 +125,268 @@
+     * @return never null, a stream that matches every A where B exists for which all the {@link BiJoiner}s are true
+     */
+    default <B> UniConstraintStream<A> ifExists(Class<B> otherClass, BiJoiner<A, B> joiner1, BiJoiner<A, B> joiner2) {
+        return ifExists(otherClass, new BiJoiner[] {joiner1, joiner2});
+    }
+
+    /**
+     * As defined by {@link #ifExists(Class, BiJoiner...)}.
+     * @param otherClass never null
+     * @param joiner1 never null
+     * @param joiner2 never null
+     * @param joiner3 never null
+     * @param <B> the type of the second matched fact
+     * @return never null, a stream that matches every A where B exists for which all the {@link BiJoiner}s are true
+     */
+    default <B> UniConstraintStream<A> ifExists(Class<B> otherClass, BiJoiner<A, B> joiner1, BiJoiner<A, B> joiner2,
+            BiJoiner<A, B> joiner3) {
+        return ifExists(otherClass, new BiJoiner[] {joiner1, joiner2, joiner3});
+    }
+
+    /**
+     * As defined by {@link #ifExists(Class, BiJoiner...)}.
+     * @param otherClass never null
+     * @param joiner1 never null
+     * @param joiner2 never null
+     * @param joiner3 never null
+     * @param joiner4 never null
+     * @param <B> the type of the second matched fact
+     * @return never null, a stream that matches every A where B exists for which all the {@link BiJoiner}s are true
+     */
+    default <B> UniConstraintStream<A> ifExists(Class<B> otherClass, BiJoiner<A, B> joiner1, BiJoiner<A, B> joiner2,
+            BiJoiner<A, B> joiner3, BiJoiner<A, B> joiner4) {
+        return ifExists(otherClass, new BiJoiner[] {joiner1, joiner2, joiner3, joiner4});
+    }
+
+    /**
+     * As defined by {@link #ifExists(Class, BiJoiner...)}.
+     * <p>
+     * This method causes <i>Unchecked generics array creation for varargs parameter</i> warnings,
+     * but we can't fix it with a {@link SafeVarargs} annotation because it's an interface method.
+     * Therefore, there are overloaded methods with up to 4 {@link BiJoiner} parameters.
+     * @param otherClass never null
+     * @param joiners never null
+     * @param <B> the type of the second matched fact
+     * @return never null, a stream that matches every A where B exists for which all the {@link BiJoiner}s are true
+     */
+    <B> UniConstraintStream<A> ifExists(Class<B> otherClass, BiJoiner<A, B>... joiners);
+
+    /**
+     * Create a new {@link UniConstraintStream} for every A, if another A exists that does not {@link #equals(Object)}
+     * the first. If multiple {@link BiJoiner}s are provided, for performance reasons, the indexing joiners must be
+     * placed before filtering joiners.
+     * @param otherClass never null
+     * @return never null, a stream that matches every A where a different A exists
+     */
+    default UniConstraintStream<A> ifExistsOther(Class<A> otherClass) {
+        return ifExists(otherClass, Joiners.filtering((a, b) -> !a.equals(b)));
```

Comment:
Due to Lambda identity, I'd implement this slight different. Let me explain:

These 2 statements use the same lambda instance:
- from(A.class).join(B.class).filter(Objects::equal)
- from(A.class).join(B.class).filter(Objects::equal)

These 2 statements do NOT use the lambda instance:
- from(C.class).join(D.class).filter((c, d) -> c.equals(d))
- from(C.class).join(D.class).filter((c, d) -> c.equals(d))

You can try this out by printing System.systemHashCode() on those BiPredicates.

Now, why do we care about that? Because of node sharing in Drools AND Bavet.
If its the same lambda instance, node sharing work guaranteed.
If it isn't, Bavet won't see it and Drools has some dirty ASM magic that might or might not recognize they are the same code.

Unfortunately, there is no Objects.notEquals in java.util.Objects.
Apache commons-lang3 does have ObjectUtils.notEqual, presuming we want to use that, I would write:

return ifExists(otherClass, Joiners.filtering(ObjectUtils::notEqual);